### ISSUE TITLE
Option for non-JSON output

### ISSUE THREAD

**Issue by stoinov** [Koenkk/zigbee2mqtt#493]
This is more of a general question: would it be feasible to add option so that each attribute to be separate entry on the MQTT path. So instead of:
`zigbee2mqtt/0x00158d0002219 {"battery":"39.00","voltage":2995,"temperature":18.87,"humidity":60.87}`
We get:
```
zigbee2mqtt/0x00158d0002219/battery 39.00
zigbee2mqtt/0x00158d0002219/voltage 2995
zigbee2mqtt/0x00158d0002219/temperature 18.87
zigbee2mqtt/0x00158d0002219/humidity 60.87
```
I am trying to setup my own automation without actually using HA, and I hit a roadblock with the [homebridge](https://github.com/oznu/docker-homebridge) setup - it does need simple strings to work, so I have to parse the JSON and republish them to make it work.

**Comment by hamstie** [comment]
I think this is a good idea:
 - .e.g  fhem can't publish and subscribe json in an uncomplicated way
 - 99% of light handling is using a single state or brightness publish

i would prefer another topic:
zigbee2mqtt/0x00158d0002219/set/state ON    # adding the key to the set command
or
zigbee2mqtt/0x00158d0002219/setkv/state ON   # for key value set

**Comment by cimba007** [comment]
I can only agree to that. Don't make the payload overly complicated by adding json.

**Comment by Koenkk** [comment]
@cimba007 JSON output is required for homeassistant integration. However I will add an option to disable this.

**Comment by rachetfoot** [comment]
I've been testing these changes and it seems to work well so far. I can make a PR with the referenced changes if it seems like a reasonable way to do it to you.
I'm not sure whether I should be converting the boolean "true" and "false" strings to on/off or open/closed. Does anyone know what's standard in MQTT?

**Comment by magcode** [comment]
That  could give you some inspiration: https://homieiot.github.io

**Comment by andreasbrett** [comment]
> That could give you some inspiration: https://homieiot.github.io

Great choice! The new MQTT binding in openHAB supports the Homie convention and would allow openHAB users to discover MQTT devices rather than having to manually enter them in openHAB.

**Comment by d33mha** [comment]
> I've been testing these changes and it seems to work well so far. I can make a PR with the referenced changes if it seems like a reasonable way to do it to you.

This would be a great feature (changing between JSON and separate topic). I am using openHAB and Zigbee2MQTT to bind Osram, Ikea and Xiaomi stuff. The new (faster and enhanced) MQTT binding is not working very well on formatting JSON for outgoing stuff.

Is this feature ready for testing?

**Comment by magcode** [comment]
I don't think a "real" implementation of the homie convention (and thus supporting the OH 2.4 MQTT auto discovery for example) has been considered/planned in this project yet. That would be quite some effort I guess.

**Comment by rachetfoot** [comment]
My change is quite naive and only handles a single level object, converting all properties of that object to topic/message pairs. It works fine for my purposes, but to be honest, I don't think it's robust enough for merging. And with what I now know about the homie convention, I think I would prefer scrapping my approach and implementing that convention instead, however that will require a much larger effort and probably also changes to the way devices are defined.

**Comment by d33mha** [comment]
> My change is quite naive and only handles a single level object, converting all properties of that object to topic/message pairs. It works fine for my purposes, but to be honest, I don't think it's robust enough for merging. And with what I now know about the homie convention, I think I would prefer scrapping my approach and implementing that convention instead, however that will require a much larger effort and probably also changes to the way devices are defined.

Is your change working as described in my assumption above? How can I implement it?

**Comment by rachetfoot** [comment]
I haven't tested, but I expect that it would publish something like this:
topic: zigbee2mqtt/mydevice/color 
message: {"r": 46,"g": 102,"b": 193}

And react to a similar message sent to zigbee2mqtt/mydevice/color/set

If you wanted to test and/or develop, you would just have to add my repo (rachetfoot/zigbee2mqtt) as a new remote where you installed it originally and then checkout the branch I made with the changes. (Also stop service before doing so and start again afterwards)

**Comment by d33mha** [comment]
> (...) I think I would prefer scrapping my approach and implementing that convention instead, however that will require a much larger effort and probably also changes to the way devices are defined.

Should we open a new issue to discuss this as a request?

**Comment by Koenkk** [comment]
What is the homie convention?

**Comment by d33mha** [comment]
> What is the homie convention?

https://homieiot.github.io

> The Homie convention defines a standardized way of how IoT devices and services announce themselves and their data on the MQTT broker.

**Comment by Koenkk** [comment]
Looks good! I think it's nice to have this in zigbee2mqtt.

**Comment by rachetfoot** [comment]
Cool, how can I help?

**Comment by Luftloch80** [comment]
@rachetfoot 
I installed your Fork and getting the mqtt in non Json format works Great
But i cannot publish anything with zigbee2mgtt/device/State/set

**Comment by rachetfoot** [comment]
Yeah, I had that issue too, and I'm not sure why because it works in the tests. However, this still works:
zigbee2mqtt/device/set

**Comment by Luftloch80** [comment]
Yes thx. This Is working

**Comment by rtreffer** [comment]
:wave: I tried the homie path - highly experimental: https://github.com/Koenkk/zigbee2mqtt/pull/855

I think zigbee2mqtt might need some changed to support plugins that act on prepared messages (before toZigbee / after fromZigbee). This would simplify custom MQTT hierarchies a lot.

**Comment by Luftloch80** [comment]
Okay. Because when i Pair the Bulbs in ioBroker they work normally

**Comment by stoinov** [comment]
@Koenkk In [the documentation](https://www.home-assistant.io/docs/mqtt/discovery/) they never mention that JSON input is required, on the contrary they are asking for separate config topic for each item you publish. The example they give is one that uses JSON, but can easily be one that you have three different state topics for each of the values.
One such example is [ESPHome](https://esphome.io/components/mqtt#using-with-home-assistant), which publish multiple items from a sensor in a separate topic as a simple number.

So even when you implement this feature, it can still work with HA given you properly setup the config topics.

**Comment by Koenkk** [comment]
@stoinov the MQTT JSON light does require this (https://www.home-assistant.io/components/light.mqtt/#json-schema) (used because this is the most complete solution: https://www.home-assistant.io/components/light.mqtt/#comparison-of-light-mqtt-schemas). 

However I completely agree that integration is also possible without JSON.

**Comment by stoinov** [comment]
@Koenkk I see. I haven't played much with lights yet, but that does makes sense.
So what is the latest status of this request? Since it got diverted with the homie discussion I am not sure if it's being worked on in parallel.

**Comment by Koenkk** [comment]
@stoinov I've made a first implementation, you can enable it with

```yaml
experimental:
  output: attribute
```

Let me know if it works out for you.

**Comment by stoinov** [comment]
@Koenkk I got:
```
/app/lib/extension/homeassistant.js:583
            throw new Error('Home Assitant integration is not possible with attribute output!');
```
but after this I saw all the payloads as simple strings in their respective topics. So great initial step!

Another thing I noticed `zigbee2mqtt/bridge/config', payload '{"log_level":"info","permit_join":false}'` - will this be stringified too?

**Comment by Koenkk** [comment]
Home assistant discover and output attribute cannot be used together (as the home assistant discovered configuration is not compatible).

We can change that too.

**Comment by stoinov** [comment]
Two questions:
1. Do you say that HA discovery is not available yet with attribute output, or that it's not technically possible?
2. How do you handle light payloads that include objects in some of their keys? Do they get further disassembled or does it happen only on the first level of the payload object?

I don't have a light to test, but from the code it seems like only the first level of attributes are reported and there is an object it's still passed as JSON. This should be easily fixable. I am just wondering whether to add another level to the topic like `topic/key/subkey` or keep them on the same level like `topic/key-subkey`.

Both have their pros and cons, and I personally use the second approach with the ESPHome reports, as it makes it easier to handle various inputs without adding too much exception logic in my scripts. Besides it all goes to DB where it's flattened anyway.

**Comment by Koenkk** [comment]
1. It's not possible because the discovered configuration expects JSON.
2. Thanks for the PR! will take a look at it ASAP

**Comment by stoinov** [comment]
I do not understand your concern about the discovered configuration. Can you elaborate?
If you mean that HA will expect the report payload to be JSON, then as long as you send numbers and not strings you'll be fine as they are considered valid JSON.
As I mentioned before - [ESPHome](https://esphome.io/) does publish single number payloads and have working HA Discovery for them.

**Comment by stoinov** [comment]
BTW I should probably update documentation too explaining the new behavior...

**Comment by Koenkk** [comment]
@stoinov because now all the attributes are reported to a different topic, so e.g. https://github.com/Koenkk/zigbee2mqtt/blob/master/lib/extension/homeassistant.js#L114 won't work anymore.

**Comment by stoinov** [comment]
I see. So we need couple of `if`s in the [discover section](https://github.com/Koenkk/zigbee2mqtt/blob/master/lib/extension/homeassistant.js#L618), to change the topics and payloads. But it will require to update to [template schema](https://www.home-assistant.io/components/light.mqtt/#template-schema) I guess. Also not sure if and why there is only xy light controlling options - is this ZigBee thing or is it just your decision?

Regardless, we'd have to convert the values to RGB with something like [this](https://gist.github.com/popcorn245/30afa0f98eea1c2fd34d).

I know it's quite the change, and most of it I think I can manage myself, but I would definitely need some help understanding some of the architectural decisions.

**Comment by Koenkk** [comment]
I don't think adding discovery configurations for both JSON and attributes output is a good idea:
- It will add a lot of extra maintenance 
- The MQTT JSON light won't work

There are also options to control via RGB: http://www.zigbee2mqtt.io/information/mqtt_topics_and_message_structure.html#zigbee2mqttdevice_idset

I think a better solution here would be to have a `json_attributes` output option, which outputs to both JSON and attributes.

**Comment by stoinov** [comment]
I agree with you that it will be a lot of extra maintenance, but only if we want to have the light as an attribute.

The optimal solution, would be to leave all light objects as JSON payloads even with attribute option. Looking at [the configurations list](https://github.com/Koenkk/zigbee2mqtt/blob/master/lib/extension/homeassistant.js#L6), only the lights section has composite payload - all other objects are single level with one entry.
So we can follow my proposal from the previous comment and just exclude light objects from this discovery changes.

We can still implement double output but I don't like this solution as our primary option since it would look and feel clumsy, especially when you try to debug MQTT.

In the RG link this explains setting up, but can you actually report in RGB or HS? There is no specific setting in the configuration list with lights, only two include XY.

**Comment by Koenkk** [comment]
That doesn't sound like a good solution me, because the light objects still have to be JSON.

The reporting is indeed in XY, as this is what zigbee uses.

**Comment by stoinov** [comment]
I am trying to setup home automation without actually using Home Assistant - just node-red parsing the MQTT payloads and inserting them into InfluxDB.
My initial setup was relying on building kind of registry using the Discovery protocol and then tagging values based on matched topics. But recently I changed it to just  using the topic structure itself as a way to tag each value so I am no longer relying on Discovery. 
Which I guess is a good argument to just leave the current state as it is and continue with something more productive :)

Going back to the request though - my whole idea is to leave the light as a JSON object and just skip the reporting of the `value_template` property in the discovery payload in all other values. This way it will still work as expected with minimal change in both JSON and attribute modes.

But as I said - it's not high priority any more and we can leave it as it is.

**Comment by olialb** [comment]
> 
> 
> @Koenkk I see. I haven't played much with lights yet, but that does makes sense.
> So what is the latest status of this request? Since it got diverted with the homie discussion I am not sure if it's being worked on in parallel.

This is really useful for openHAB integration. Do you plan to make this from experimental to "advanced"

**Comment by Koenkk** [comment]
@olialb this feature will stay in for sure, not sure if it's well tested enough atm though.

**Comment by andreasbrett** [comment]
> @olialb this feature will stay in for sure, not sure if it's well tested enough atm though.

I've been using it for months now without issues.

**Comment by stoinov** [comment]
I've been using it almost a year, so it's stable. But I do not have much experience with setting it up in different environments, so not sure if we have some edge cases that could break it, so any feedback is welcomed.

I use this exclusively, without HA obviously. Any other cases?

**Comment by olialb** [comment]
I am using it with openHab and it runs stable. Make the setup much more easy....

**Comment by krzyk** [comment]
I used this feature for read-pnly attributes, but now I have an IKEA light bulb and would like to control it without using json.

Is it possible?
What are the topics?

```
zigbee2mqtt/[FRIENDLY_NAME]/brightness
zigbee2mqtt/[FRIENDLY_NAME]/color/r
zigbee2mqtt/[FRIENDLY_NAME]/color/g
zigbee2mqtt/[FRIENDLY_NAME]/color/b
```

How about transitions, if it is no longer a single json(/transaction) it would be hard to do transition to given state, so how it is handled right now?

**Comment by kianusch** [comment]
**zigbee2mqtt/[FRIENDLY_NAME]/set/brightness**

seems to work.

**zigbee2mqtt/[FRIENDLY_NAME]/set**

works with with **ON|OFF** payload.

**Comment by nicomania** [comment]
Hello all,
red through this and I am not sure how to enable the feature of non-Json output. I have put 
```
  publish_as_json: false
  publish_as_key_topics: true
```
in my `configuration.yaml`. I tried it at `mqtt:` and `advanced:` but non worked. I still get json mqtt messages only. As it is not mentioned in documentation, can somebody please help me to set this up?
Thank you!

**Comment by nicomania** [comment]
figured it out. For everybody facing the same:
put `output: 'attribute'` under `advanced:` of yaml.
Greetings!
